api/lib/domain/services/user-service.js

```diff
@@ -13,204 +11,49 @@ const competenceRepository = require('../../../lib/infrastructure/repositories/c
 const knowledgeElementRepository = require('../../../lib/infrastructure/repositories/knowledge-element-repository');
 const courseRepository = require('../../../lib/infrastructure/repositories/course-repository');
 
-async function _findCorrectAnswersByAssessments(assessments) {
-  const answersByAssessmentsPromises = assessments.map((assessment) =>
-    answerRepository.findCorrectAnswersByAssessmentId(assessment.id));
+const getProfileToCertifyV1 = _.partial(_getProfileToCertify, _getUserCompetencesAndCorrectlyAnsweredChallengesIdsV1);
+const getProfileToCertifyV2 = _.partial(_getProfileToCertify, _getUserCompetencesAndCorrectlyAnsweredChallengesIdsV2);
 
-  const answersByAssessments = await Promise.all(answersByAssessmentsPromises);
-
-  return _.flatten(answersByAssessments);
-}
-
-function _getUserCompetenceByChallengeCompetenceId(userCompetences, challenge) {
-  return challenge ? userCompetences.find((userCompetence) => userCompetence.id === challenge.competenceId) : null;
-}
-
-function _findChallengeBySkill(challenges, skill) {
-  return _(challenges).filter((challenge) => {
-    return challenge.hasSkill(skill) && challenge.isPublished();
-  }).value();
-}
-
-function _skillHasAtLeastOneChallengeInTheReferentiel(skill, challenges) {
-  const challengesBySkill = _findChallengeBySkill(challenges, skill);
-  return challengesBySkill.length > 0;
+function _getProfileToCertify(getUserCompetencesAndAnswersFn, userId, limitDate) {
+  return getUserCompetencesAndAnswersFn({ userId, limitDate })
+    .then(UserCompetence.getCertificationInputDatas(challengeRepository.list))
+    .then(UserCompetence.updateWithChallenges);
 }
 
-function _createUserCompetencesV1({ allCompetences, allAdaptativeCourses, userLastAssessments }) {
-  return allCompetences.map((competence) => {
-    const userCompetence = new UserCompetence(competence);
-    const currentCourse = allAdaptativeCourses.find((course) => course.competences[0] === userCompetence.id);
-    const assessment = userLastAssessments.find((assessment) => currentCourse.id === assessment.courseId);
-    if (assessment) {
-      userCompetence.pixScore = assessment.getPixScore();
-      userCompetence.estimatedLevel = assessment.getLevel();
-    } else {
-      userCompetence.pixScore = 0;
-      userCompetence.estimatedLevel = 0;
-    }
-    return userCompetence;
+async function _getUserCompetencesAndCorrectlyAnsweredChallengesIdsV1({ userId, limitDate }) {
+  return UserCompetence.getWithCorrectlyAnsweredChallengesIdsV1({
+    coursesFetcher: courseRepository.getAdaptiveCourses,
+    competencesFetcher: _.memoize(competenceRepository.list),
+    lastAssessmentsFetcher: _.memoize(_.partial(assessmentRepository.findLastCompletedAssessmentsForEachCoursesByUser, userId, limitDate)),
+    correctAnswersFetcher: answerRepository.findCorrectAnswersByAssessmentId,
```

Comment:
Probablement la ligne la plus exotique du fichier mais en fait elle n'a rien de compliqué. L'idée sous-jacente notamment dans ma compréhension de la clean archi qu'on essaye de mettre en place est l'inversion de contrôle, c'est à dire, les `model` qui définissent eux-même l'interface dont ils ont besoin. C'est l'idée derrière tous les `xxxFetcher()` qu'on retrouve. 

La pire situation c'est la suivante : 
```
// model.js
const repo = require('...');
```
Puisque dans ce cas, la dépendance n'étant même pas injecté, on peut difficilement la substituer. Le modèle connait tout le repo et son implem (méthodes, interfaces, signatures...) 
Un peu mieux c'est d'injecter la dépendance, ce qu'on fait pour les `usecase` de façon automatiser:
```
// model.js
function myModelMethod( repository ) { ... } 
```
Mais le modèle a encore connaissance de tout le repo ! Du coup on peut imaginer que le modèle lui même définisse l'interface qu'il souhaite, qu'il la nomme comme il veut et faire confiance à l'appelant pour lui fournir ce dont il a besoin : 
```
// model.js
function myModelMethod( { userSaver } ) { ... } 
// service.js
model.myModelMethod({ userSaver: userRepository })
```
Là ça commence à ressembler vraiment à de l'inversion de contrôle mais y'a encore un pb. Le modèle doit apprendre à utiliser les méthodes, alors dans l'absolue, ce qui l'intéresse vraiment, c'est juste d'aller chercher des user. Il se fiche de savoir comment. Il veut être capable d'aller chercher des objets de _quelque part_ si possible sans savoir comment (= n'avoir aucun argument). La solution c'est donc de laisser le service lui passer directement la fonction avec les arguments déjà en place. C'est ce que fait `_.partial` 
```
// model.js
function myModelMethod( { userSaver } ) { 
    userSaver(); // Youhou! 
} 
// service.js
model.myModelMethod({ userFetcher: _.partial(userRepository.save, userId) })
```
Et là on a bien un métier qui s'est affranchit de tout ce qui est superflu. Au prix d'un peu de technicité, on a un métier épuré réduit à son essence, complètement isolé de tout le reste. On pourrait se demander si le simple fait d'aller chercher les données ne devrait pas être dans une coucher et n'avoir que des méthodes synchrone. Peut-être c'est ça la différence entre `usercase` et `entity` après-tout. En tout cas je n'ai pas été jusque là.

Reply:
Concernant le `memoize`, son utilisation permet là aussi de simplifier _grandement_ en splittant en deux des fonctions dont la cohésion n'est qu'accidentelle, à savoir pour des raisons de performance. On est souvent tenté, lorsqu'on a des traitement **_asynchrones non-parallelisables_** à fetcher le premier batch de données, faire un premier traitement dessus synchrone, et une deuxième requête asynchrone, qui fera elle même son traitement une fois les données reçu. 

Dans l'absolu, ces deux choses pourraient être séparées, car elles n'ont pas forcément grand chose à voir entre elles, si ce n'est qu'elles partagent le résultat d'une opération asynchrone. Du coup on est tenté de les rassembler dans des fonctions qui prennent souvent des `And` dans leurs nom qui traduit uniquement le fait que pour éviter de faire deux-allers retour à la base ou sur le réseau, on les met côte à côte et on a des fonctions avec des interfaces d'abstraction différentes. 

Au lieu d'avoir `function (userId, assessmentId)` on a des `function (userId, assessments)` où l'assessment a déjà été fetché par la fonction d'avant, mais où le userId va être utilisé pour faire un `await` directement. 

Utiliser `_.memoize` permet de s'affranchir de ce problème et de découper ses fonction exactement comme on le souhaite, comme si ce qui était asynchrone ne l'était pas. Il faut juste faire attention à ce ne `memoize` que des variables crées localement qui seront rapidement détruites (pas de `memoize` sur les singletons des `require` par exemple)